subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -609,24 +614,40 @@ private ResourceLock getOrCreateProjectLock(Project project) {
         return workerLeaseService.getProjectLock(gradlePath, projectPath);
     }
 
-
     private boolean canRunWithCurrentlyExecutedTasks(TaskInfo taskInfo) {
         TaskInternal task = taskInfo.getTask();
 
         Pair<TaskInternal, String> overlap = firstTaskWithOverlappingOutput(task);
-        if (overlap == null) {
-            return true;
-        } else {
+        if (overlap != null) {
             LOGGER.info("Cannot execute task {} in parallel with task {} due to overlapping output: {}", task.getPath(), overlap.left.getPath(), overlap.right);
+            return false;
         }
 
-        return false;
+        Set<TaskInfo> runningTaskInfos = CollectionUtils.collect(runningTasks, new Transformer<TaskInfo, TaskInternal>() {
+            @Override
+            public TaskInfo transform(TaskInternal task) {
+                return executionPlan.get(task);
+            }
+        });
+        overlap = firstTaskWithDestroyedIntermediateInput(taskInfo, runningTaskInfos);
+        if (overlap != null) {
+            LOGGER.info("Cannot execute task {} in parallel with task {} due to overlapping input/destroy: {}", task.getPath(), overlap.left.getPath(), overlap.right);
+            return false;
+        }
+
+        overlap = firstTaskWithDestroyedIntermediateInput(taskInfo, executionQueue);
```

Comment:
I've implemented what you've described such that we only cache the outputs once the task has started and we track the consumer relationships and remove them once the consumer is complete.

I've also changed it such that it's an error to specify both destroyables and inputs/output files on a task.  We'll tackle the inputs + destroyables case in a follow up story.